proposals/0000-term-lookup-rules.md

```diff
@@ -0,0 +1,200 @@
+
+  data T = T
+
+  p = ProxyV T
+  ```
+  Does `T` on the last line refer to the data constructor `T` or the type constructor `T`?
+
+  Here's a more real-world example, let's look at `sizeOf` in `Foreign.Storable`:
+  ```haskell
+  sizeOf :: Storable a => a -> Int
+  ```
+  This function takes a `Storable a => a` value as an argument when in reality it doesn't really need an actual value, it only cares about its type. Currently, to use it we have to provide it with a value, and the most convenient way to do it is to use `undefined`:
+  ```
+  sizeOf (undefined :: MyType)
+  ```
+  Ideally, we want to be able to do something like this instead:
+  ```haskell
+  sizeOf :: forall a -> Storable a => Int
+
+  sizeOf MyType
+
+  -- or even
+
+  f :: forall a. something
+  f = ... sizeOf a ...
+  ```
+  Of course right now we have an option to use `@` syntax to do that:
+  ```haskell
+  sizeOf :: forall a. Storable a => Int
+
+  sizeOf @MyType
+  ```
+  But that argument is implicit and not required. If the user forgets to specify it, GHC will try to infer it which will lead to bad error messages like "could not deduce" that does not really hint to what the problem is and how to fix it.
+
+These problems could be solved by changing the identifier lookup rules in a non-intrusive way: whenever we lookup an identifier at the term level we can first search for it in the data namespace, and if it wasn't found there, we could search in other namespaces as a fall-back. To keep compatibility with modules that use punning, a new syntax could be introduced allowing to import only terms or only types from a module.
+
+## Proposed Change Specification
+
+* The name lookup rules are extended: when looking up term-level identifier fails, look for a type-level identifier as a fall-back.
+* The lookup logic extended to look up in type namespace on terms.
+* To disambiguate the namespaces of identifiers from modules that use punning, the `-XExplicitNamespaces` extension is extended with new syntax:
+  ```
+  impdecl   -> import [qualified] modid [as modid] [impspec]
+             |
+  ```
+  is changed to
+  ```
+  impdecl   -> import [qualified] modid [namespace] [as modid] [impspec]
+             |
+
+  namespace -> data
+             | type
+  ```
+   * With `data` specified in the import, only identifiers belonging to the data namespace will be brought into the scope.
+   * With `type` specified in the import, only identifiers belonging to the type namespace will be brought into the scope.
+* A new module `Data.Type` is introduced which defines unambiguous type definitions for built-in syntax constructors: `List`, `Tuple` and `Unit`:
+  ```haskell
+  type List = []
+  type Unit = ()
+  type Tuple2 = (,)
+  type Tuple3 = (,,)
+  type Tuple4 = (,,,)
+  {- ... -}
+
+  type family Tuple (xs :: [Type]) :: Type where
+    Tuple [] = Unit
+    Tuple [a] = GHC.Tuple.Unit a
+    Tuple [a,b] = Tuple2 a b
+    Tuple [a,b,c] = Tuple3 a b c
+    {- ... -}
+  ```
```

Comment:
What are your thoughts on @Ericson2314's ideas in https://github.com/ghc-proposals/ghc-proposals/pull/270#issuecomment-537974057 as well as @int-index's idea of including `~` as a built-in type?

Reply:
While I'm all for it, I feel like it is a little bit out of the scope of this proposal.

Reply:
True, but the proposal process is costly (time-wise), so we could try to batch these changes in a single proposal as an optimization. I would also include @goldfirere's "further" section from https://github.com/ghc-proposals/ghc-proposals/pull/270#issuecomment-536954938

It's OK to create a separate section in the proposal for this (e.g. "Extended Proposed Change Specification"), I've seen it done before.

Reply:
(I should really have put this review on the line mentioning `Data.BuiltInTypes` rather than `Data.Type`, but anyway)

> While I'm all for it

Does that refer to both points? https://github.com/ghc-proposals/ghc-proposals/pull/270#issuecomment-537974057 makes it sound like `Data.BuiltInTypes` shouldn't be used at all, and the built-in types instead added to already existing modules `GHC.Tuple` and `GHC.List` – _if_ that is the eventual goal, it would probably make more sense to handle it that way immediately.

Reply:
Sorry, I was referring to `a ~ b` for an operator. As far as module structure goes I agree that it makes more sense to have list and tuple constructors be in separate modules.

I also agree that it would be nice to have them exported by `Prelude` there's a drawback to this: every time you'd want to use lists on the type level you'd have to enable `-XNoImplicitPrelude` and import prelude explicitly hiding the constructors. I'd imagine would affect a lot of code (anything that uses vinyl-records, for example, or servant) by forcing users to import Prelude explicitly.